### PR TITLE
apis: add KubernetesApply, ImageMap types

### PR DESCRIPTION
**Pull request by nicks**
Hello @milas, @landism,

Please review the following commits I made in branch nicks/k8s:

1f60a199a3ac2d3a5bbdc00dfc3450c30d660f7c (2021-06-09 12:43:20 -0400)
apis: add KubernetesApply, ImageMap types

Code review reminders, by giving a LGTM you attest that:

* Commits are adequately tested
* Code is easy to understand and conforms to style guides
* Incomplete code is marked with TODOs
* Code is suitably instrumented with logging and metrics

### CODE REVIEW

**Comment by nicks** [comment]
this doesn't define any fields yet, the important part is the two comments that describe how these objects behave.

File: pkg/apis/core/v1alpha1/register.go [hunk 1]
```diff
@@ -48,12 +48,14 @@ func AllResourceObjects() []resource.Object {
 		&Session{},
 		&FileWatch{},
 		&Cmd{},
+		//&KubernetesApply{},
```
Comment (milas) [inline pkg/apis/core/v1alpha1/register.go:51]: Did you want to hide these from the API for now?

**Review by milas — APPROVED** [review]

File: pkg/apis/core/v1alpha1/imagemap_types.go [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// ImageMap expresses the mapping from an image reference to a real, pushed
+// image in an image registry that a container runtime can access.
+//
+// Another way to think about the ImageRef is that ImageMapSpec is a mutable
```
Comment (landism) [inline pkg/apis/core/v1alpha1/imagemap_types.go:38]:
```suggestion
// Another way to think about the ImageMap is that ImageMapSpec is a mutable
```

guessing this was a missed rename, or else not sure what it's referring to

**Review by landism — APPROVED** [review]
neat!

just a couple nits

File: pkg/apis/core/v1alpha1/kubernetesapply_types.go [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KubernetesApply specifies a blob of YAML to apply, and a set of ImageMaps
+// that the YAML depends on.
+//
+// The KubernetesApply controller will resolve the ImageMaps into immutable image
+// references, modify the YAML so that it can pull those images, then apply
```
[the remark this answers is not in the document]
Comment (nicks) [reply pkg/apis/core/v1alpha1/kubernetesapply_types.go:39]: done

File: pkg/apis/core/v1alpha1/imagemap_types.go (hunk 2 above)
Comment (nicks) [reply pkg/apis/core/v1alpha1/imagemap_types.go:38]: done

File: pkg/apis/core/v1alpha1/register.go (hunk 1 above)
Comment (nicks) [reply pkg/apis/core/v1alpha1/register.go:51]: ya, will uncomment once we need controllers :+1:

**Pull request merged**